### DISCUSSION TITLE
File Upload Difficulties

### DISCUSSION THREAD

**Discussion by Bill-Lathrop** [fastapi/fastapi#7074] [Questions] [answered]
Hello - I am not having any luck with file upload to FastAPI - I am trying to load JPEG images.  I have reviewed the other github, stackoverflow comments and none of the solutions work. 

I believe the issue is with the front-end - and I am able to get it to work with the swagger \docs interface.  From my front-end I have tried submitting directly from a form as well as formData via XMLHttpRequest and fetch (would prefer either of the latter).  I am using plain javascript. 

I am sure I am doing something wrong - I just don't know what it is.  Any advice would be deeply appreciated - I am 20+hours deep into trying to solve this one.  **Thanks in advance!**

**Here is the swagger curl:**
``curl -X POST "http://localhost:8000/staff/upload_image/" -H  "accept: application/json" -H  "Content-Type: multipart/form-data" -F "image=@_WGL1281.jpg;type=image/jpeg"
``

**Server Reply:  POST http://localhost:8000/staff/upload_image/ 422 (Unprocessable Entity)**

**Installed Versions:**
```
fastapi - 0.54.1
starlette - 0.13.2
pydantic - 1.5.1
python - 3.832
python-multipart - 0.0.5

```
**Server side code snippets:**

```
from fastapi import APIRouter, Request, Depends, File, UploadFile

@router.post("/staff/upload_image/")
async def upload_image(image: UploadFile = File(...)):
	print(image.filename)
	return {"filename": image.filename}

```

**client side:**
```
async function imageupload(){
            let imageinput = document.getElementById('testimage');
            let formData = new FormData();
            formData.append('imagefile', imageinput.files[0]);

            let response = await fetch('/staff/upload_image/', {
                method: 'POST',
                body: formData
            })

            await console.log(response);

```

**Comment by Mause** [comment]
Can you try with the latest version of fastapi, and give us the error from the server logs?

**Comment by Bill-Lathrop** [comment]
I have updated FastAPI and dependencies to the following:
```
fastapi - 0.62.0
starlette - 0.13.6
```

Server log using my front end code:
```
←[32mINFO←[0m:     127.0.0.1:54609 - "←[1mGET /staff_main HTTP/1.1←[0m" ←[32m200 OK←[0m
←[32mINFO←[0m:     127.0.0.1:54609 - "←[1mGET /favicon.ico HTTP/1.1←[0m" ←[31m404 Not Found←[0m
←[32mINFO←[0m:     127.0.0.1:54609 - "←[1mPOST /staff/upload_image/ HTTP/1.1←[0m" ←[31m422 Unprocessable Entity←[0m
```
And with the swagger docs interface:
```
←[32mINFO←[0m:     127.0.0.1:54610 - "←[1mGET /docs HTTP/1.1←[0m" ←[32m200 OK←[0m
←[32mINFO←[0m:     127.0.0.1:54610 - "←[1mGET /openapi.json HTTP/1.1←[0m" ←[32m200 OK←[0m
_WGL0071.jpg
←[32mINFO←[0m:     127.0.0.1:54611 - "←[1mPOST /staff/upload_image/ HTTP/1.1←[0m" ←[32m200 OK←[0m
```
where the _WGL0071.jpg is a response to print the file name on the server.

**Comment by Bill-Lathrop** [comment]
So the above results are using fetch.  When I switch to XMLHttpRequest - I get a different message in my console.log - in case this helps.  

`{"detail":[{"loc":["body","image"],"msg":"field required","type":"value_error.missing"}]}
`

**Comment by Bill-Lathrop** [comment]
One more thing - if this works in swagger \docs - do we know how its front-end works?  I've tried to figure that out - but the code is not easy to examine.

**Comment by ycd** [comment]
@Bill-Lathrop have you tried setting a `Header`? 


```js
async function imageupload(){
            let imageinput = document.getElementById('testimage');
            let formData = new FormData();
            formData.append('imagefile', imageinput.files[0]);

            let response = await fetch('/staff/upload_image/', {
                method: 'POST',
                body: formData,
                headers: {"Content-Type": "image/jpeg"}
            })

            await console.log(response);
```

**Comment by Bill-Lathrop** [comment]
@ycd - yes I've tried setting headers - as the FastAPI docs mention, to "multipart/form-data".  Many of the issues reported elsewhere say not to do this - and it didn't work for me either.  I just tried the image/jpeg as you suggest - no improvement.  

Results with fetch: ←[1mPOST /staff/upload_image/ HTTP/1.1←[0m" ←[31m400 Bad Request←[0m

Results with XLMHttpRequest: {"detail":"There was an error parsing the body"} is the response.  

Thanks for taking a look though!

**Comment by dstlny** [comment]
I've personally had issues uploading documents using FastAPI too on my personal VueJS+FastAPI project. What solved it for me, was setting the Content-Type header to "undefined". Bare in mind, i am using Axios, but i feel it's probably similar using fetch?

**Comment by Bill-Lathrop** [comment]
@dstlny - thanks so much for the idea.  I tried it with both fetch and XMLHttpRequest - neither worked.  Error with Request was {"detail":[{"loc":["body","image"],"msg":"field required","type":"value_error.missing"}]}

Ugh!  My project is a platform for creating and congregating local news - the ability to load images and other files is critical.  I am able to do this fine with Flask - but really like the speed and structure of FastAPI.  Starting to take a look at Node JS.  As I am the only developer - it might be nice to work exclusively in JS - though I really like Python.

**Comment by dstlny** [comment]
> @dstlny - thanks so much for the idea. I tried it with both fetch and XMLHttpRequest - neither worked. Error with Request was {"detail":[{"loc":["body","image"],"msg":"field required","type":"value_error.missing"}]}
> 
> Ugh! My project is a platform for creating and congregating local news - the ability to load images and other files is critical. I am able to do this fine with Flask - but really like the speed and structure of FastAPI. Starting to take a look at Node JS. As I am the only developer - it might be nice to work exclusively in JS - though I really like Python.

Out of curiosity, what happens if you remove the "headers" entry from the request? (to force the browser to determine the correct headers)

**Comment by Jonatha-Varjao** [comment]
@Bill-Lathrop hmmm I think there's a little typo 
the API form data expects a key named image, but on your frontend side, you're sending it as 'imageFile'.
Can you check this for me?
Hope it helps

**Comment by Bill-Lathrop** [comment]
@dstlny - I tried removing headers - and no effect.  Same errors.  
@Jonatha-Varjao - also no impact.  I have seen the formData append with examples like "myFile", etc - so I assumed it did not matter what it was called.  But - as you suggest - I tried "image" and "file" and neither worked. 

Thanks though!

**Comment by dstlny** [comment]
@Bill-Lathrop  - are you able to inspect the request that your browser is sending to check that you are able to see the form data? Could you also check that the Content-Type looks something like: "multipart/form-data; boundary=----WebKitFormBoundaryM2zKSP2Vjo1PFMWy" - where Boundary is something randomly generated (this should be automatically figured out by the browser)?

**Comment by Bill-Lathrop** [comment, marked as answer]
@dstlny - Sorry for the slow reply - got sidetracked by other things yesterday. 

Ran the process again this morning - and able to analyze the request payload:

```
------WebKitFormBoundaryZh8xzFE9hB5ZTlYN
Content-Disposition: form-data; name="image"; filename="_WGL0315.jpg"
Content-Type: image/jpeg
------WebKitFormBoundaryZh8xzFE9hB5ZTlYN--
```

Well - after another attempt I may have just had some success - but am not sure why.  I removed the section on setting the content header and reran with a result of status=200 with a filename in response.  I now have to verify I can save the file on the server - but this is the closest to success I've had.  

Thanks for keeping me company on this journey!  I'll follow-up when I can successfully save the file.

**Comment by Bill-Lathrop** [comment]
@dstlny - Luke - thanks for your help.  I've been able to save the uploaded file - so I guess this issue is resolved.  I am not sure what the issue finally was - but glad to have it figured out and working in at least one configuration. 

Now the tough decision - I started looking at Node JS a couple days ago and have set up a server and started playing with it - and I really like it.  

Best of luck to you in your career and life journey!

**Comment by michaeltoohig** [comment]
I just wanted to say for others facing the same problem as I have, I found the issue to be the fetch api itself. I tested it across a couple browsers but always got the ```{detail: "There was an error parsing the body"}``` response. If I remove the `multipart/form-data` headers I get a 422 error.

I tried using axios and it worked as expected so something about fetch API is the issue IMO.

**Comment by Bill-Lathrop** [comment]
So - I had a similar problem working with Node.js/Express.  I was able to resolve it there somewhat more quickly - mostly due to experience gained with FastAPI.  I think an important step is to ensure you put the encoding type into either the `<form>` or the `<input>`.

` enctype="multipart/form-data"`

This seems to have been the change that made it work.  I'm not sure why and it seems counterintuitive since I was pulling the data from the form into a formData object via javascript - not posting the form directly.  But - this is the combination that seems to work and I am using fetch.  

I don't think I was able to get fetch to work with FastAPI, so I probably agree with @michaeltoohig 

Good luck!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:

**Comment by Ibrahim-abdelaziz** [comment]
@tiangolo Could you please guide me to fix this issue, when i upload file or image local it save correct in folder but when i upload on development server after deploying the file or image not save in folder that created in project directory, i tried many times but i can't fix this issue , thanks in advanced


@attachments_router.post("/attachments")
async def create_attachments(
    files: List[UploadFile] = File(...),
    db: Session = Depends(get_db)
):
    for file in files:
        with open("media/" + file.filename, "wb") as image:
            shutil.copyfileobj(file.file, image)
        file_path = str("media/" + file.filename)
        data = {}
        data[state] = state_id
        data["attachment"] = file_path
        obj_in = AttachmentsCreateIn.parse_obj(data)
        result = crud.attachments.create(db_session=db, obj_in=obj_in)
   return {"Message": "File Uploaded Successfully"}

**Comment by includeamin** [comment]
> file_path = str("media/" + file.filename)

``` python
import os
cwd = os.getcwd()

file_path = os.path.join(cwd,"media",file.filename)
```

**Comment by 3cham** [comment]
> @dstlny - I tried removing headers - and no effect. Same errors.
> @Jonatha-Varjao - also no impact. I have seen the formData append with examples like "myFile", etc - so I assumed it did not matter what it was called. But - as you suggest - I tried "image" and "file" and neither worked.
> 
> Thanks though!

Actually your form should use the field name as "image" since you used it as expected parameter in your upload_image method.

For those who aren't aware of this, in your html file just define the field name as same as the variable in the handling method, e.g. 

```html
<input type="file" name="my_field_name"/>
```

then in server side:
```python
async def process_upload(my_field_name: UploadFile = File(...)):
     do_some_thing_with(my_field_name)
```

**Comment by galdreiman** [comment]
Thank you all for the ideas.
As I recently encountered the same issue (frontend: React app and backend: FastAPI) I want to share the final solution:
In my frontend app I have an API request wrapper that by default defines the request headers, include `Content-Type: application/json`.
Removing the header and letting the browser to figure out the content type solved the problem!

**Comment by ChaturvedRaj** [comment]
> > @dstlny - I tried removing headers - and no effect. Same errors.
> > @Jonatha-Varjao - also no impact. I have seen the formData append with examples like "myFile", etc - so I assumed it did not matter what it was called. But - as you suggest - I tried "image" and "file" and neither worked.
> > Thanks though!
> 
> Actually your form should use the field name as "image" since you used it as expected parameter in your upload_image method.
> 
> For those who aren't aware of this, in your html file just define the field name as same as the variable in the handling method, e.g.
> 
> ```
> <input type="file" name="my_field_name"/>
> ```
> 
> then in server side:
> 
> ```python
> async def process_upload(my_field_name: UploadFile = File(...)):
>      do_some_thing_with(my_field_name)
> ```

I was facing the same issue and Solution provided by @3cham resolved it...thanks for the help everyone..you guys saved me lot of trouble....uploaded file name and input tag name should be same

**Comment by rajfnu** [comment]
I had the same/similar issue and fixed it by changing the CURL request as below (hope it helps others):

`
curl -X 'POST' \
  'http://localhost:8000/api/v1/upload/user/{user_id}' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'fileobject=@test.png;type=image/png'
`

what is important here is to notice that the Python code below has "fileobject" name same as "-F 'fileobject' in the curl command. Also, -F in the curl command has 'type' in it.
`
@router.post("/upload/user/{user_id}", status_code=200, description="Upload image to S3")
async def upload(fileobject: UploadFile = File(...)):
`

**Comment by lifeconnectng** [comment]
I had the same/similar issue. i am using React for frontend and Fastapi for backend, axios for API calls.
   <input type="file"onChange={saveFile} name="file" accept=".jpeg, .png, .jpg" enctype="multipart/form-data"/>

export const uploadImages = async (data) => {console.log(data.file)
  return await axios.post(
    `${process.env.REACT_APP_BASE_URL}images/imageUpload`,config
  );
};

and the fastapi part: @router.post('/imageUpload')
async def imageUpload(
    file: UploadFile = File(...), db:Session =Depends(get_db)):
    return 'Image Successfully Uploaded'

throwing a 422 error: {"detail":[{"loc":["body","file"],"msg":"field required","type":"value_error.missing"}]}

**Comment by gigeos** [comment]
I use axios but I tried with fetch api and it works like this with React Native and FastAPI to upload a file to S3

FetchAPI (React Native)
```
    const url = `.../aws/file`;
    const bodyFormData = new FormData();
    bodyFormData.append('fileb', {
      uri: photo.uri,
      name: 'pictureName.jpg',
      type: 'image/jpeg',
    });
      const response = await fetch(url, {
        body: bodyFormData,
        method: 'POST',
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });
```

FastAPI
```
@aws_controller.post('/file/community', status_code=200, description="Upload filee to S3")
async def post_aws_file(fileb: UploadFile = File(None)):
    return aws_upload_file_bucket(
        fileb, 'bucket-name')
        
        ...
        
def aws_upload_file_bucket(file_object, bucket_name:
     filename = file_object.filename
        split_file_name = os.path.splitext(filename)
        file_extension = split_file_name[1]
        data = file_object.file._file
        key = uuid()+file_extension
        s3_resource.Bucket(name=bucket_name).put_object(Key=filename, Body=data)

```

**Comment by aryan10behal** [comment]
I faced similar issue. I was trying to send a csv file from frontend to backend. My issue got resolved by removing CONTENT-TYPE from header in my request.

My frontend in ReactJS

```
    const handleSubmissionDelete = async() => {
        if(!isFilePickedDelete) return;
        const formData = new FormData();
        formData.append("file", selectedFileDelete);
        const requestOptions = {
            method: "POST",
            headers: {
            Authorization: "Bearer " + token,
            },
            body: formData,
          };   
          const response = await fetch(env.BACKEND + `/deleteLightsFile`, requestOptions);
          const data = await response.json();
      
          if (!response.ok) {
            setErrorMessage(data.detail);
          } else {
            setLights([])
          }
    }
```

And my backend looked like

```
@app.post("/deleteLightsFile")
async def deleteLightsFile(file: UploadFile = File(...), user: _schemas.User = Depends(_services.get_current_user)):
    
    authenticate_user(user.email)
    df = pd.read_csv(file.file, dtype={'Unique Pole No.':str, 'Wattage':str})
    lampposts = []
     .....
    return {'response': "successfully deleted all csv poles"}
```

**Comment by franchesoni** [comment]
Could you guess what's wrong?

backend:
```python
@app.post("/upload")
async def post_img(file: UploadFile):
    logger.info("new image")
    contents = await file.read()
    Img = Image.open(io.BytesIO(contents))
    iaseg.set_PIL_Image_and_reset(Img)
    return {"status": "ok"}
```

html:
```html
    <input type="file" id="fileInput" accept=".jpeg">
```

frontend:
```typescript
const fileInput = document.getElementById('fileInput') as HTMLInputElement;
fileInput.addEventListener('change', async () => {
  const files = fileInput.files
  if (files && files.length > 0) {
  const file = files[0];
  const reader = new FileReader();
  reader.readAsArrayBuffer(file);

  reader.onload = async () => {
    const imgArrayBuffer = reader.result as ArrayBuffer;
    console.log(imgArrayBuffer)
    console.log("uploading image")

    const response = await fetch(`http://${IPaddress}:${port}/upload`, {
      method: 'POST',
      // headers: {
      //   'Content-Type': 'image/jpeg'
      //   'Content-Type': 'application/octet-stream'
      // },
      body: imgArrayBuffer
    });
    if (response.ok) {
      const imgBlob = new Blob([imgArrayBuffer], {type: 'image/jpeg'});
      img.src = URL.createObjectURL(imgBlob);  // resets and handles image
    }
    else {
      console.log("error when uploading image")
    }
  }
  };
}
);
```

but I get **422 (Unprocessable Entity)**

**Reply by yyamano** [reply]
Try to send FormData object instead of imgArrayBuffer.

**Comment by felixweingartner** [comment]
Using fastapi + vite react js axios the below strategy worked for me->
Set these three headers sending your file in formdata:
if (options.body instanceof FormData) {
      options.headers['Accept'] = '*/*';
      options.headers['Content-Type'] = 'multipart/form-data';
      options.headers['Enctype'] = "multipart/form-data";
}

**Comment by Jayanth-MKV** [comment]
Solved !!
The below code works - main thing is don't forget to include the filename in FormData options.

> import * as FormData from 'form-data';

> import { Buffer } from 'node:buffer';

       const formData = new FormData();
       const buf = await data.arrayBuffer(); // here data is Blob - Blob { size: 97728, type: 'audio/wav' }
       // convert blob to buffer as formdata doesnot accept it directly.
       formData.append("file", Buffer.from(buf),{
       filename:"audio.wav",
        });
       const apiResult = await axios.post(ENDPOINT,formData);

**Comment by zlaast** [comment]
I actually ran into the problem of getting **422 (Unprocessable Entity)** and after a few hours, I got something that worked, though I don't know if it will be helpful. Wherever you see `yourIdentifier`  just change it to your identifier (you should probably avoid using dashes or underscores here). You'll also need to change the endpoints of course.

**Frontend: HTML** 
```html
<form @input="uploadFunction" enctype="multipart/form-data">
    <input type="file" accept="image/*" capture="user" id="yourIdentifier" name="yourIdentifier" />
</form>
```

Note I'm using VueJS here, hence the `@input`, so you'll have to do some modifications for whatever framework you're using.
The `@input` just fires the `uploadFunction` function when `<input>` changes.

**Frontend: Typescript**
```javascript
async function uploadFunction(e: Event)
{
    const yourIdentifier= new FormData(e.currentTarget as HTMLFormElement)
    
    fetch(`http://YOURURL/ENDPOINT`, {
        method: 'POST',
        headers: { 'accept': 'application/json' },
        body: yourIdentifier
    });
}
```

Above, I'm using `e.currentTarget` for `FormData` but you can get your files however you need.

**Backend**
```python
@app.post("/ENDPOINT")
async def file_upload(yourIdentifier: UploadFile):
    print(yourIdentifier.filename)
    return {"filename": yourIdentifier.filename}
```

In the example above, I have 7 instances of `yourIdentifier`, so change those and it should work (mandatory "but it worked on *my* computer).
